### ISSUE TITLE
scaleBar remains hidden until the map's visible region changes

### ISSUE THREAD

**Issue by mmertsock** [mapbox/mapbox-gl-native#11302]
**Platform: iOS**
**Mapbox SDK version: 3.7.5**

### Steps to trigger behavior

- Create an MGLMapView and immediately set `mapView.scaleBar.isHidden = false`.

I'm creating the MGLMapView, and adding it to the view hierarchy, programmatically rather than using a storyboard — not sure if that matters for reproducing this.

### Expected behavior

I expected the scale bar to be visible as soon as the view appears. (This is what I'm trying to achieve in my app: a scale bar that's visible 100% of the time.)

### Actual behavior

The scale bar remains hidden initially. But as soon as the user pans or zooms the map, the scale bar does become.

**Comment by friedbunny** [comment]
Thanks for reporting this issue, @mmertsock.

The underlying issue here is that the scale bar view is driven solely by camera update events in the map view — the scale bar view can’t update itself, so setting `mapView.scaleBar.isHidden = false` (even with a custom setter) doesn’t provide enough information to draw.

I see two potential ways to fix this:

- Add `MGLMapView.showsScale` and have its setter control the visibility of the scale bar view. This would replace direct `mapView.scaleBar.isHidden` as the primary way to enable the scale bar.
- Have the scale bar view hold a weak reference to the map view, from which it can glean the necessary information to draw itself.

/cc @frederoni @1ec5

**Comment by lloydsheng** [comment]
@mmertsock 

After initialization, the scaleBar view's width was 0. that's why you can't see it. The actual width will be calculated while map cameraI changing that mostly triggered by user's pan or zoom gestures.
If you want the scaleBar shows at the same time as mapview. here is a workaround might help you:
```swift
let centerCoordinate = CLLocationCoordinate2D(latitude: 38.90, longitude: -77.03)
let mapView = MGLMapView(frame: view.bounds, styleURL: url)
mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
mapView.setCenter(centerCoordinate, zoomLevel: 14, animated: false)
mapView.delegate = self
mapView.scaleBar.isHidden = false
// make the scalebar shows 
let metersPerPoint = mapView.metersPerPoint(atLatitude: centerCoordinate.latitude);
mapView.scaleBar.setValue(metersPerPoint, forKey: "metersPerPoint")
```
By the way, I was wondering if the behavior is by design.

**Comment by 1ec5** [comment]
> The underlying issue here is that the scale bar view is driven solely by camera update events in the map view

Why doesn’t this camera change within `-[MGLMapView commonInit]` lead to a call to `-cameraIsChanging` that updates the scale bar?

https://github.com/mapbox/mapbox-gl-native/blob/848d7dff4e682c3d2485366f95649a5de4b1bd75/platform/ios/src/MGLMapView.mm#L566

**Comment by friedbunny** [comment]
That’s a clever workaround, thanks @lloydsheng.

As for the design of showing/hiding the scale bar automatically, we decided to punt that to the developer. Whereas something like MapKit will fade the scale bar in/out based on user input, our scale bar should always be potentially visible (within a range) and it’s up to the developer to decide if/when to hide it.

@1ec5 My guess is that we don’t get `cameraIsChanging` because `jumpTo` is zero duration. I’m not familiar with the core implementation behind this, but that’s what we document publicly:

https://github.com/mapbox/mapbox-gl-native/blob/9a30e252b53b654d9b756f839bd33f2ff6db1d91/platform/ios/src/MGLMapViewDelegate.h#L102-L105

**Comment by 1ec5** [comment]
That’s a good point. Anything triggered by `-mapViewRegionIsChanging:` should also be triggered by `-mapView:regionDidChange:`, because the final frame of an animation (or the only frame for a non-animated camera change) causes only `-mapView:regionDidChange:` to get called, not `-mapViewRegionIsChanging:`.

**Comment by friedbunny** [comment]
This issue was fixed by #11335, which will be included in `ios-v4.0.0`. Thanks again for reporting it, @mmertsock.
